test/functional/ui/highlight_spec.lua

```diff
@@ -650,3 +650,296 @@ describe("'listchars' highlight", function()
     ]])
   end)
 end)
+
+describe("'winhighlight' highlight", function()
+  local screen
+
+  before_each(function()
+    clear()
+    screen = Screen.new(20,8)
+    screen:attach()
+    screen:set_default_attr_ids({
+      [0] = {bold=true, foreground=Screen.colors.Blue},
+      [1] = {background = Screen.colors.DarkBlue},
+      [2] = {background = Screen.colors.DarkBlue, bold = true, foreground = Screen.colors.Blue1},
+      [3] = {bold = true, reverse = true},
+      [4] = {reverse = true},
+      [5] = {background = Screen.colors.DarkGreen},
+      [6] = {background = Screen.colors.DarkGreen, bold = true, foreground = Screen.colors.Blue1},
+      [7] = {background = Screen.colors.DarkMagenta},
+      [8] = {background = Screen.colors.DarkMagenta, bold = true, foreground = Screen.colors.Blue1},
+      [9] = {foreground = Screen.colors.Brown},
+      [10] = {foreground = Screen.colors.Brown, background = Screen.colors.DarkBlue},
+    })
+    command("hi Background1 guibg=DarkBlue")
+    command("hi Background2 guibg=DarkGreen")
+  end)
+
+  it('works', function()
+    insert("aa")
+    command("split")
+    command("set winhl=Normal:Background1")
+    screen:expect([[
+      {1:a^a                  }|
+      {2:~                   }|
+      {2:~                   }|
+      {3:[No Name] [+]       }|
+      aa                  |
+      {0:~                   }|
+      {4:[No Name] [+]       }|
+                          |
+    ]])
+
+    command("enew")
+    screen:expect([[
+      {1:^                    }|
+      {2:~                   }|
+      {2:~                   }|
+      {3:[No Name]           }|
+      aa                  |
+      {0:~                   }|
+      {4:[No Name] [+]       }|
+                          |
+    ]])
+  end)
+
+  it('handles invalid values', function()
+    command("set winhl=Normal:Background1")
+    screen:expect([[
+      {1:^                    }|
+      {2:~                   }|
+      {2:~                   }|
+      {2:~                   }|
+      {2:~                   }|
+      {2:~                   }|
+      {2:~                   }|
+                          |
+    ]])
+
+    eq('Vim(set):E474: Invalid argument: winhl=xxx:yyy',
+       exc_exec("set winhl=xxx:yyy"))
+    eq('Normal:Background1', eval('&winhl'))
+    screen:expect([[
+      {1:^                    }|
+      {2:~                   }|
+      {2:~                   }|
+      {2:~                   }|
+      {2:~                   }|
+      {2:~                   }|
+      {2:~                   }|
+                          |
+    ]])
+  end)
+
+
+  it('works local to the buffer', function()
```

Comment:
That was decided by bram long ago. "window option" means global=per window, local=per buffer.